Added two first use cases in form of full web stack integration tests.

alien-mcl opened the pull request:
Adding necessary infrastructure for simulated web client behavior so we can add use-cases related tests and data sets. This should also give a good playground for reference data sets for tests - this discussion was already raised some time ago.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hydracg/heracles.ts/3)
<!-- Reviewable:end -->

alien-mcl commented:
I've noticed that for some situations JSON-LD framing has some issues. I've raised an issue [here](https://github.com/digitalbazaar/jsonld.js/issues/191) related.

tpluscode commented:
I'd be more concerned that in the playground [you link to](https://json-ld.org/playground/#startTab=tab-framed&json-ld=%7B%22%40context%22%3A%7B%22%40context%22%3A%7B%22%40vocab%22%3A%22http%3A%2F%2Fwww.w3.org%2Fns%2Fhydra%2Fcore%23%22%2C%22hydra%22%3A%22http%3A%2F%2Fwww.w3.org%2Fns%2Fhydra%2Fcore%23%22%2C%22entrypoint%22%3A%7B%22%40id%22%3A%22hydra%3Aentrypoint%22%2C%22%40type%22%3A%22%40id%22%7D%7D%7D%2C%22%40id%22%3A%22%2Fapi%3Fdocumentation%22%2C%22entrypoint%22%3A%7B%22%40id%22%3A%22%2Fapi%22%2C%22%40type%22%3A%22EntryPoint%22%7D%7D&frame=%7B%22%40context%22%3A%7B%22iri%22%3A%22%40id%22%2C%22isA%22%3A%7B%22%40id%22%3A%22%40type%22%2C%22%40container%22%3A%22%40set%22%7D%2C%22%40vocab%22%3A%22http%3A%2F%2Fwww.w3.org%2Fns%2Fhydra%2Fcore%23%22%2C%22hydra%22%3A%22http%3A%2F%2Fwww.w3.org%2Fns%2Fhydra%2Fcore%23%22%2C%22operations%22%3A%22hydra%3Aoperation%22%2C%22entryPoint%22%3A%7B%22%40id%22%3A%22hydra%3Aentrypoint%22%7D%7D%7D&context=%7B%7D) the `/api` resource appears twice in JSON graph. 

I think that flattened is a more robust form for processing. Especially that we are transforming the descriptions to a separate representation, right?

alien-mcl commented:
Well, the `/api` part is something you won't have control over - it's up to the server how it'll return resource addresses.
As for the flattening - I'm not sure what did you have in mind. Was it about framing? In general, this is the only API method that allows to link real objects instead of resource references in the graph which should serve the purpose of making a separate representation.

tpluscode commented:
Look again at the framed JSON-LD `@graph`:

``` json
[
    {
      "iri": "../api",
      "isA": "EntryPoint"
    },
    {
      "iri": "../api?documentation",
      "entryPoint": {
        "iri": "../api",
        "isA": "EntryPoint"
      }
    }
  ]
```

You get `"iri": "../api"` twice. Which means you will get two JS objects. This is inevitable with graph represented as tree. I'm pretty sure it was a problem for me before.

lanthaler commented:
Reviewed 26 of 26 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[doc/classes/_src_apidocumentation_.apidocumentation.md, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8xnw1Ck43Usf8isyG:-Ko8xnw2aIESAM1t8DzH:b-i48d26) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_apidocumentation_.apidocumentation.md#L39)):*
> ```Markdown
> *Implementation of [IApiDocumentation](../interfaces/_src_datamodel_iapidocumentation_.iapidocumentation.md).[client](../interfaces/_src_datamodel_iapidocumentation_.iapidocumentation.md#client)*
> 
> *Defined in [src/ApiDocumentation.ts:19](https://github.com/alien-mcl/Heracles.ts/blob/63ad02f/src/ApiDocumentation.ts#L19)*
> ```

Could we make this point to master instead so that we won't have as many diffs going forward?

---

*[doc/classes/_src_hydraclient_.hydraclient.md, line 263 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8yoziqE346Wu9Pvfu:-Ko8yoziqE346Wu9Pvfv:b-tuob0p) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_hydraclient_.hydraclient.md#L263)):*
> ```Markdown
> | Name  | Type                | Default | Description  |
> | ------ | ------------------- | ------------ | ------------ |
> | urlOrResource  | string⎮object | - | Url or a {@link IResource} carrying an Iri of the resource to be obtained. |
> ```

Nit: Is it expected that this isn't properly aligned?

---

*[integration-tests/HydraClient.spec.ts, line 10 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zK684M_s3ExQH9km:-Ko8zK684M_s3ExQH9kn:bnfma4i) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L10)):*
> ```XML
>     });
> 
>     describe("while browsing a website", function() {
> ```

a website -> the test website

---

*[integration-tests/HydraClient.spec.ts, line 20 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zNEK6VP3MSczgz15:-Ko8zNEK6VP3MSczgz16:b-mb4zvf) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L20)):*
> ```XML
>             }));
> 
>             it("should obtain two hypermedia", function() {
> ```

two hypermedia controls

---

*[integration-tests/HydraClient.spec.ts, line 25 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zUZ7JEetzfGm3PJ2:-Ko8zUZ7JEetzfGm3PJ3:b5md8hy) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L25)):*
> ```XML
> 
>             it("should obtain a schema:CreateAction operation", function() {
>                 expect(this.entryPoint.hypermedia[0].isA).toBe("Operation");
> ```

I think we should make this assertion (and the ones below) order independent

---

*[integration-tests/server/server.ts, line 17 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91NgbvO8PlVDuCUe4:-Ko91NgbvO8PlVDuCUe5:b4tf6n4) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L17)):*
> ```XML
> require.extensions[".headers"] = textFileLoader;
> require.extensions[".txt"] = textFileLoader;
> require.extensions[".jsonld"] = jsonFileLoader;
> ```

Wouldn't it be simpler to treat everything as text and use fs.readFileSync directly below?

---

*[integration-tests/server/server.ts, line 35 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko9-sADVNVBMkKcuA0b:-Ko9-sADVNVBMkKcuA0c:br8tcc7) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L35)):*
> ```XML
>     });
> 
> function sendHeaders(path: string, response: express.Response, isBodyLoaded: boolean): boolean
> ```

setHeaders might be a more appropriate name for this method. I'd also rename isBodyLoaded to isJsonLd as that's the only thing that flag is used for.

---

*[integration-tests/server/api/documentation.jsonld, line 3 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91oWwbeIUGRoQ6ist:-Ko91oWwbeIUGRoQ6isu:befiiir) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/api/documentation.jsonld#L3)):*
> ```JSONLD
> {
>     "@context": "/api/context.jsonld",
>     "@id": "/api?documentation",
> ```

This should be /api/documentation

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-Ko93XJdk94rIdRt1tIF:bvn96ul) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
> ```XML
>      * @returns {Promise<IWebResource>}
>      */
>     public async getResource(urlOrResource: string | { iri: string }): Promise<IWebResource>
> ```

Hmm... I'm not sure we should use urlOrResource throughout the codebase.. but if we do, we should probably define a type alias for it.

---

*[src/HydraClient.ts, line 120 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko92N7uDs0Z1HniifqX:-Ko92N7uDs0Z1HniifqY:b-fetli5) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L120)):*
> ```XML
>         }
> 
>         return (!result[1].match(/^[a-z][a-z0-9+\-.]*:/) ? jsonld.prependBase(url.match(/^[a-z][a-z0-9+\-.]*:\/\/[^/]+/)[0], result[1]) : result[1]);
> ```

Probably we should work with expanded (and perhaps) flattened JSON-LD internally. Such kind of operations open a whole can of worms

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko95PGVXlixJsGAwyLv:-Ko95PGVXlixJsGAwyLw:bgh1grx) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L26)):*
> ```XML
>     {
>         let payload = await response.json();
>         let flattened = await jsonLd.flatten(payload, null, { base: response.url.match(/^[a-z][a-z0-9+\-.]*:\/\/[^/]+/)[0] });
> ```

I don't think I understand this. Why do you call response.url.match here?

---

*[tests/HydraClient.spec.ts, line 92 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko94joCRJXRRWB3d7sj:-Ko94joCRJXRRWB3d7sk:b84sdr3) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/tests/HydraClient.spec.ts#L92)):*
> ```XML
>             describe("and that documentation is provided in an unsupported format", function() {
>                 beforeEach(function() {
>                     let apiDocumentationUrl = `${this.expectedUrl}api/documentation`;
> ```

We should probably rename expectedUrl to baseUrl given how you use it here

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko97g0lJVatt2PKZh04:-Ko97g0mAnE8-Ft5UWOs:b-1lk3uv) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
> ```XML
>             let result = await this.hypermediaProcessor.process(this.response, true);
> 
>             expect(result.hypermedia).toEqual(hypermediaJsonLd);
> ```

I think it would be easier to read and understand these tests if these files would be inlined here.

Btw. I'm still quite skeptical about this "hypermedia" removal

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
The example you posted in the [framing bug](https://github.com/digitalbazaar/jsonld.js/issues/191) is invalid JSON-LD. You have a `@context` nested inside a `@context`. Also, redefining keywords doesn't work, you can only alias them. So You can't change the behavior of `@type` this way. There exists a [`compactArrays`](https://www.w3.org/TR/json-ld-api/#widl-JsonLdOptions-compactArrays) option in all conformant JSON-LD processors that you could leverage to achieve the same though.

Btw. if you run into issues like this, you can cross-check the result of the jsonld.js playground with mine that is available at (the UI should probably be modernized at some point :smile:) http://www.markus-lanthaler.com/jsonld/playground/

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-:-Ko988uA1Gp1e0wGzOWM:b-6jd929)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
As for the `/api` - I'm not sure how expanded/flattened form would behave here. I don't see any option fro framing to get rid of these.

As for the framing bug - that nested `@context` originally was a link to a context file, which I couldn't pass to the playground - I just simply replaced it with the source, didn't notice it creates nested one. But back to the framing - it's somehow troublesome as documentation is somehow difficult to read and the examples are quite poor. Intent was to define a container for `@type` to be consistent, but I didn't want to use global options as I acknowledge some places could benefit from the non-array behavior.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[doc/classes/_src_apidocumentation_.apidocumentation.md, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8xnw1Ck43Usf8isyG:-Ko9Uc90381wM_PCnLt2:b-lkoty4) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_apidocumentation_.apidocumentation.md#L39)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Could we make this point to master instead so that we won't have as many diffs going forward?
</blockquote></details>

OK, didn't notice it

---

*[doc/classes/_src_hydraclient_.hydraclient.md, line 263 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8yoziqE346Wu9Pvfu:-Ko9UjjVRKBJda8kLrxF:b3ibr35) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_hydraclient_.hydraclient.md#L263)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Nit: Is it expected that this isn't properly aligned?
</blockquote></details>

Not sure - this is auto-generated, maybe the template has issues. need to check

---

*[integration-tests/HydraClient.spec.ts, line 10 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zK684M_s3ExQH9km:-Ko9VmNsVLXwJUVA3hWI:bb74ce2) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L10)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

a website -> the test website
</blockquote></details>

Sure

---

*[integration-tests/HydraClient.spec.ts, line 20 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zNEK6VP3MSczgz15:-Ko9Vo37q4ItT2yMi7eH:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L20)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

two hypermedia controls
</blockquote></details>

OK

---

*[integration-tests/HydraClient.spec.ts, line 25 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zUZ7JEetzfGm3PJ2:-Ko9UrJ9D6XdqE0gXCOb:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L25)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I think we should make this assertion (and the ones below) order independent
</blockquote></details>

OK

---

*[integration-tests/server/server.ts, line 17 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91NgbvO8PlVDuCUe4:-Ko9UxfD8GNeG4oFyx_m:bohhtki) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L17)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Wouldn't it be simpler to treat everything as text and use fs.readFileSync directly below?
</blockquote></details>

But we don't want to treat JSON-LD files as text - we want it as JSON.

---

*[integration-tests/server/server.ts, line 35 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko9-sADVNVBMkKcuA0b:-Ko9VpP8WfnnQZfUfKSY:bb74ce2) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L35)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

setHeaders might be a more appropriate name for this method. I'd also rename isBodyLoaded to isJsonLd as that's the only thing that flag is used for.
</blockquote></details>

Sure

---

*[integration-tests/server/api/documentation.jsonld, line 3 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91oWwbeIUGRoQ6ist:-Ko9V8_XGe2rNX0XdlGz:b-fhv7u7) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/api/documentation.jsonld#L3)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

This should be /api/documentation
</blockquote></details>

Yep, copy-pasted from use-cases

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-Ko9ROgvLx11zWAcNagt:b-vz7fmp) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Hmm... I'm not sure we should use urlOrResource throughout the codebase.. but if we do, we should probably define a type alias for it.
</blockquote></details>

Normally I'd leave only string, but framing won't change a resource to a literal.

---

*[src/HydraClient.ts, line 120 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko92N7uDs0Z1HniifqX:-Ko9Rpa_FwLsuXtvafbq:b-jkjy88) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L120)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Probably we should work with expanded (and perhaps) flattened JSON-LD internally. Such kind of operations open a whole can of worms
</blockquote></details>

My previous experience led me to separate data model. Using expanded form for some cases was really inconvenient (sub-classing, etc.).
This also allows to inject into that model controls i.e. from headers (especially for non-RDF payloads)

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko95PGVXlixJsGAwyLv:-Ko9SVaXwQhY2zBj55_M:b97cnsb) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L26)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I don't think I understand this. Why do you call response.url.match here?
</blockquote></details>

We need to give a base Url for JSON-LD operations in case some relative ones are inside. It won't get the document's Url as we own the response, which carries the requested resource's Url (after all possible redirections).

---

*[tests/HydraClient.spec.ts, line 92 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko94joCRJXRRWB3d7sj:-Ko9V_OcPIyW4xjpB5AO:bb74ce2) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/tests/HydraClient.spec.ts#L92)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

We should probably rename expectedUrl to baseUrl given how you use it here
</blockquote></details>

Sure

---

*[tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts, line 39 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko97g0lJVatt2PKZh04:-Ko9SyKUzPTpUzgnZSAK:b163crs) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/tests/DataModel/JsonLd/JsonLdMetadataProvider.spec.ts#L39)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

I think it would be easier to read and understand these tests if these files would be inlined here.

Btw. I'm still quite skeptical about this "hypermedia" removal
</blockquote></details>

Indeed, removal should be an option. In this test I wanted to cover that scenario.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-:-Ko9T9d63dKdTKBVCO4V:b-8cmun1)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Yeah, please keep in mind that framing hasn't been standardized (yet) so the various JSON-LD processors don't always behave the same way. I wonder whether we need to frame it at all or whether we should just work with flattened JSON-LD even the triples directly internally? For the latter, we may be able to leverage [some of the work of the RDFJS CG](https://github.com/rdfjs/representation-task-force/blob/master/interface-spec.md).

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 10 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zK684M_s3ExQH9km:-KoJMWr8WW_umapkBP4m:bgb6cd5) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L10)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Sure
</blockquote></details>

Could you please push the code updates with the comments, then we can start closing some of these discussions and focus on the other ones

---

*[integration-tests/server/server.ts, line 17 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91NgbvO8PlVDuCUe4:-KoJN5I3KoYZ7lXlTdFu:brf6ygs) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L17)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

But we don't want to treat JSON-LD files as text - we want it as JSON.
</blockquote></details>

Sure, in the client.. but this is the (dumb) server, right? 

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-KoJNM11wdDKCddkQEnm:bg6s6he) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Normally I'd leave only string, but framing won't change a resource to a literal.
</blockquote></details>

In other words, you don't want the user to get a resource's URL before calling this method?

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko95PGVXlixJsGAwyLv:-KoJOBd-Q_qieuCle5p2:b4uiwc8) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L26)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

We need to give a base Url for JSON-LD operations in case some relative ones are inside. It won't get the document's Url as we own the response, which carries the requested resource's Url (after all possible redirections).
</blockquote></details>

My question was, why do we need the call to match? Won't it always be set? Would it ever be set to something other than an absolute URL or null/undefined?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-:-KoJKomtUY_mVYfm9TJO:b-qokqcv)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
As I wrote in one of the comments - my experiences with implementing Hydra clients earlier lead me to conclusion that raw triples are no good. For some simple scenarios - fine, but any trace of i.e. subclassing introduces complex code forks.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[integration-tests/HydraClient.spec.ts, line 10 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8zK684M_s3ExQH9km:-KoOGFxj9A-KxprTkfyh:b-aerqpp) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/HydraClient.spec.ts#L10)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Could you please push the code updates with the comments, then we can start closing some of these discussions and focus on the other ones
</blockquote></details>

I'll do it this evening

---

*[integration-tests/server/server.ts, line 17 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko91NgbvO8PlVDuCUe4:-KoO8byH3Nm_2S5MK4iZ:b-8l9c6g) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/integration-tests/server/server.ts#L17)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Sure, in the client.. but this is the (dumb) server, right? 
</blockquote></details>

This dumb server will need some extra logic (i.e. for filtering) sooner than later, but OK - I'll change it.

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-KoO9NoE7e6hgrNTYemA:b-g83vw6) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

In other words, you don't want the user to get a resource's URL before calling this method?
</blockquote></details>

I'm not sure I understand correctly.  In general - I'd prefer to have it as string only as we're interested only in the url. Unfortunate thing is that with framing I'm unable to convert i.e. entry point from a resource `{ "@id": "some url" }` to a string `"some url"`. With this approach method can utilize both.
I'll introduce a separate type that will hold an IRI alone as an alias.

---

*[src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts, line 26 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko95PGVXlixJsGAwyLv:-KoOCHz-mCV5pddfZSKO:bxuch0) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/DataModel/JsonLd/JsonLdHypermediaProcessor.ts#L26)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

My question was, why do we need the call to match? Won't it always be set? Would it ever be set to something other than an absolute URL or null/undefined?
</blockquote></details>

Ahh, you're right - we shall use full Url in case the context is relative to document rather than to the site.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-:-KoOEU7gbB__peuKkt03:b-j49hff)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
I think we should move this to a separate discussion and proceed with this PR for now.

---

Reviewed 42 of 42 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[package.json, line 10 at r2](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQkmdLGQhvfKPHFZTD:-KoQkmdLGQhvfKPHFZTE:bad372d) ([raw file](https://github.com/hydracg/heracles.ts/blob/7df376de1d9e08264260dd4ea376710f17a463d9/package.json#L10)):*
> ```JSON
>     "serve": "ts-node integration-tests/server/server.ts",
>     "docs-html": "node node_modules/typedoc/bin/typedoc --out doc --excludeExternals --tsconfig tsconfig.docs.json --sourcefile-url-prefix \"https:\/\/github.com\/\/HydraCG\/Heracles.ts\/blob\/master\/\" src",
>     "docs-md": "node node_modules/typedoc/bin/typedoc --out doc --theme markdown --excludeExternals --tsconfig tsconfig.docs.json --sourcefile-url-prefix \"https:\/\/github.com\/\/HydraCG\/Heracles.ts\/blob\/master\/\" src"
> ```

There are two slashes after above: `github.com\/\/HydraCG -> github.com\/HydraCG`

---

*[doc/classes/_src_hydraclient_.hydraclient.md, line 263 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8yoziqE346Wu9Pvfu:-KoQneygV3oRzlYPSUZM:b-n6m49o) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_hydraclient_.hydraclient.md#L263)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

Not sure - this is auto-generated, maybe the template has issues. need to check
</blockquote></details>

OK, let's leave this for later.

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-KoQmfwOYY9iF-h5SLEk:b-ewemer) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I'm not sure I understand correctly.  In general - I'd prefer to have it as string only as we're interested only in the url. Unfortunate thing is that with framing I'm unable to convert i.e. entry point from a resource `{ "@id": "some url" }` to a string `"some url"`. With this approach method can utilize both.
I'll introduce a separate type that will hold an IRI alone as an alias.
</blockquote></details>

Let's discuss this later and proceed with this PR first

---

*[src/HydraClient.ts, line 120 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko92N7uDs0Z1HniifqX:-KoQmzk6xRRxszvjwWZm:bwk8ul) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L120)):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

My previous experience led me to separate data model. Using expanded form for some cases was really inconvenient (sub-classing, etc.).
This also allows to inject into that model controls i.e. from headers (especially for non-RDF payloads)
</blockquote></details>

OK, let's defer this discussion too for now.

---

*[src/HydraClient.ts, line 73 at r2](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQmS09V2hmC1PFs-db:-KoQmS09V2hmC1PFs-dc:bwqsehs) ([raw file](https://github.com/hydracg/heracles.ts/blob/7df376de1d9e08264260dd4ea376710f17a463d9/src/HydraClient.ts#L73)):*
> ```XML
>     /**
>      * Obtains a representation of a resource.
>      * @param urlOrResource {string | IResource} Url or a {@link IHydraResource} carrying an Iri of the resource to be obtained.
> ```

Comment is inconsistent with param definition I**Hydra**Resource

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-:-KoQnoffprPcbDP-77wt:bjzq8o3)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQobP2ybEXpcrf74Bp:-KoQobP2ybEXpcrf74Bq:b-lcvrpf) (no related file):*
Btw. after you fixed the last two nits, could you please rebase this PR on master head so that Travis runs. I'll then merge this unless someone objects in the meantime.

Thanks a lot for your work Karol

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 7 of 7 files at r3, 2 of 4 files at r4.
Review status: 48 of 50 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQobP2ybEXpcrf74Bp:-KossuefLNZRvGpGE93-:b-nr0bdg) (no related file):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Btw. after you fixed the last two nits, could you please rebase this PR on master head so that Travis runs. I'll then merge this unless someone objects in the meantime.

Thanks a lot for your work Karol
</blockquote></details>

Try to merge in https://github.com/HydraCG/Heracles.ts/pull/9. It should resolve the issues with the tests.

---

*[karma.conf.js, line 11 at r3](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KossbpyPkOMquvdI-_z:-KossbpyPkOMquvdI-a-:b-h2j7do) ([raw file](https://github.com/hydracg/heracles.ts/blob/31c0c461504bb1ae04be41b9895a6b33f691152b/karma.conf.js#L11)):*
> ```JavaScript
>             { pattern: "integration-tests/**/*.spec.ts", included: true }
>         ],
>         exclude: (process.env.TRAVIS ? ["jsonld-request", "server", "integration-tests"] : ["jsonld-request", "server"]),
> ```

Is this actually necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 1 of 4 files at r4.
Review status: 49 of 50 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->

alien-mcl commented:
Review status: 49 of 50 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQobP2ybEXpcrf74Bp:-Kox9B8bRySU0UXSI1e-:b-kbelkn) (no related file):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Try to merge in https://github.com/HydraCG/Heracles.ts/pull/9. It should resolve the issues with the tests.
</blockquote></details>

I was able to overcome issues with tests. I had to replace npm-run-all with concurrently as it allows to exit gracefully in case a one of the tasks exits. Now I run HTTP server and karma runner in single run mode. Once it finishes whole tasks exits.
I've also merged with master, thus travis seems to like it now.

---

*[karma.conf.js, line 11 at r3](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KossbpyPkOMquvdI-_z:-Kox8N1sX3oMFUZBKmSc:b-ezn80d) ([raw file](https://github.com/hydracg/heracles.ts/blob/31c0c461504bb1ae04be41b9895a6b33f691152b/karma.conf.js#L11)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Is this actually necessary?
</blockquote></details>

Nope, I was able to get rid of it.

---

*[package.json, line 10 at r2](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQkmdLGQhvfKPHFZTD:-Kox8WlgpsneK8FJFuDX:b-886gsx) ([raw file](https://github.com/hydracg/heracles.ts/blob/7df376de1d9e08264260dd4ea376710f17a463d9/package.json#L10)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

There are two slashes after above: `github.com\/\/HydraCG -> github.com\/HydraCG`
</blockquote></details>

Fixed

---

*[doc/classes/_src_hydraclient_.hydraclient.md, line 263 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko8yoziqE346Wu9Pvfu:-Kox8btAyU5TWVm5n9y-:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/doc/classes/_src_hydraclient_.hydraclient.md#L263)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

OK, let's leave this for later.
</blockquote></details>

OK

---

*[src/HydraClient.ts, line 82 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko93XJdk94rIdRt1tIE:-Kox940HDiOgZezsn9mM:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L82)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Let's discuss this later and proceed with this PR first
</blockquote></details>

OK

---

*[src/HydraClient.ts, line 120 at r1](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-Ko92N7uDs0Z1HniifqX:-Kox96-7-gM8gHYq3UiG:b32w0x) ([raw file](https://github.com/hydracg/heracles.ts/blob/b79ce7ad78bd50944bf02dc46bfaa0c12e5f00dc/src/HydraClient.ts#L120)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

OK, let's defer this discussion too for now.
</blockquote></details>

OK

---

*[src/HydraClient.ts, line 73 at r2](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQmS09V2hmC1PFs-db:-Kox91S-4KEXf9pOo2EL:b-886gsx) ([raw file](https://github.com/hydracg/heracles.ts/blob/7df376de1d9e08264260dd4ea376710f17a463d9/src/HydraClient.ts#L73)):*
<details><summary><i>Previously, lanthaler (Markus Lanthaler) wrote…</i></summary><blockquote>

Comment is inconsistent with param definition I**Hydra**Resource
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->

lanthaler commented:
Reviewed 1 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/hydracg/heracles.ts/3#-KoQobP2ybEXpcrf74Bp:-KoxDNoVvMjn3RcJEwVP:b-1n1okb) (no related file):*
<details><summary><i>Previously, alien-mcl (Karol Szczepański) wrote…</i></summary><blockquote>

I was able to overcome issues with tests. I had to replace npm-run-all with concurrently as it allows to exit gracefully in case a one of the tasks exits. Now I run HTTP server and karma runner in single run mode. Once it finishes whole tasks exits.
I've also merged with master, thus travis seems to like it now.
</blockquote></details>

OK. I'll merge this now and rebase the other PR. I think it would be better to have a single process and let Karma start express then introducing something like concurrently. We can discuss that in the other PR. Great work Karol!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hydracg/heracles.ts/3)*
<!-- Sent from Reviewable.io -->
